Optimize matching ids

todolson opened the pull request:
This is initial work on optimizing `BibDB::matchingIDs`. The goal is to limit the amount of work done by the browse handler when generating browse entries that have large number of hits.

Long term, we want to refactor PHP code to always construct links out of the browse list using the heading, so `matchingIDs()` will become obsolete.

TODO
- [ ] Simplify logic around `id` and `extra` fields:
- - [x] merge id field with extra fields
- - [x] Refactor `BrowseItem` and `BrowseList` so we can add directly to response
- - [ ] add `fields` parameter to query as a wrapper around `extras`
- - [ ] migrate all code to use `fields` internally
- [ ] get hit count directly from Lucene, not by counting all of the matching IDs.
- [ ] Implement `maxBibListSize` limit
- [x] ~~Implement retrieveBibId logic~~ - obsolete

todolson commented:
The most recent changes just adds "ids" to the list of fields to be retrieved. I've not yet figured out how to do this as a wrapper. I assume the goal is to have a run-time switch that changes behavior. If compile-time is sufficient, then the fiddling with the list of fields is done.

demiankatz commented:
This certainly looks like a logical step forward -- it not only makes the code more flexible but also makes it more concise. I think we had talked about making this a run-time-configurable behavior. Maybe the thing to do is to trace back and figure out where the extras parameter passed to matchingIDs is coming from, since it's probably in the config parsing where it makes the most sense to inject the "ids" as part of the field list, rather than hard coding it into the method where it currently is... and of course it also seems like renaming both the matchingIDs function and the extras parameter would make a lot of sense after these changes are in place. Obviously that's a significant BC break, but my feeling is that it's probably not a big deal to do so as long as the external-facing API is unchanged.

todolson commented:
extras is first seen by `BrowseRequestHandler.handleRequestBody()`, it comes is as a parameter in the HTTP request, see `BrowseRequestHandler.java` line 868. So it is somehow created by the PHP code.

It's not obvious to me how to cleanly add "ids" to extras as a run-time configuration. Maybe add a switch in the solrconfig.xml entry for each browse index? The `BrowseSource`s are created in `BrowseRequestHandler.init()`, so the switch could then be a toggle in that class, and each `BrowseSource` decides whether it adds "ids" to extras? The `retrieveBibId` switch might already be poised to serve that function if we set the default to true.

todolson commented:
That seems reasonable at first reading. The two parameters can co-exist until the PHP code uses only one, then in a year remove "extras" entirely. I'll look at this next time I can turn attention to the project.

Also, I had good info on how to do what we with the SimpleCollector, courtesy of the Lucene java-users list. That's really the heart of the matter. But I would like to get the parameters set up well if I can make the time.

demiankatz commented:
Sounds like a plan! Let me know if I can do anything more to help.

todolson commented:
After digging into the code a little more, I am reminded that we now also have a `retrieveBibId` setting that can be set in solrconfig.xml on a per-browse-index basis. That is a boolean which triggers adding ids if they are not explicitly asked for by the calling code. Does the `fields` idea make that obsolete, or is there a need for both?

demiankatz commented:
I think the fields idea may make that obsolete. If we can control the behavior through the HTTP request rather than the Solr configuration, that makes everything simpler and more flexible and seems preferable.

demiankatz commented:
@todolson, just a heads-up for whenever you get around to working on this: you'll probably want to merge the latest master, which has been updated to support Solr 7. The Travis build process now tests against the current VuFind master branch rather than a fixed build, to ensure that we're always testing the latest and greatest.

todolson commented:
Have rebased with master and tests run to success locally. Here's hoping for the same.
I hope to do some real work on this before the New Year.

todolson commented:
The problem with the checks right now is that `/home/travis/build/vufind-org/vufind-browse-handler/build/deps` has not been created before this statement is run to define the property  `classpathProp`:
```xml
<pathconvert property="classpathProp" refid="classpath"/>
```
It's only used for debugging, so I will just comment it out.

demiankatz commented:
Thanks again for the progress, @todolson -- as always, let me know if you need my input on anything.

todolson commented:
As I'm looking through the code, it occurs to me that for searching the authority a biblio cores we could consider switching from using the Solr classes like SolrIndexSearcher to access the cores directly to using SolrJ classes to access those cores via HTTP. 

The good part is we would get rid of the need to reference count the cores and such. The  bad parts are (1) that there may be more serializing and deserializing of the responses, and (2) we would be reimplementing something that already works. I don't know which is likely to be more future-proof, the current strategy of direct index access via the Solr classes, or access via the SolrJ client.

My inclination is to not take on this reimplementation at this time, but to leave a record of the idea in case it seems appropriate at some point in the future.

demiankatz commented:
@todolson, as you say, this idea is probably beyond the scope of the current PR, but it does seem like it could be worthwhile -- not only for the reasons you mention, but also because it might improve compatibility with sharded/cloud instances of Solr. Feel free to open a JIRA ticket so we don't lose track of this (and so we can discuss further in its own context).

EreMaijala commented:
@demiankatz I'm afraid the whole mechanism would need an overhaul to be compatible with SolrCloud or a sharded index. Since it requires, as far as I remember, that the SQLite database be present in any node handling the request, you'd have to build the browse index on each node or add a mechanism that ships the index from the building node to the others.

todolson commented:
@EreMaijala Yes, that's right. We run a master-slave configuration and arrange to copy SQLite indexes from master to slave after they are generated. It's not pretty.
One possibility could be to move the indexes from SQLite to MySQL, but then there is the complication of direct MySQL access from a Solr extension. For Blacklight, I think Naomi Dushay figured out a way to do alphabetical browse all inside of Solr. But any of these are serious undertakings.

demiankatz commented:
@EreMaijala and @todolson, this is true -- the indexing process would have to change as well as the browse handler, and it would be non-trivial work. But nonetheless, using the HTTP API instead of direct index access would at least remove one of the obstacles to such a change.

todolson commented:
@demiankatz @marktriggs I've been looking at how to build unit tests to actually examine the content of results returned by the browse handler. Am using the SolrJ library as a client and an `EmbeddedSolrServer` to access the core. When `BrowseRequestHandler#handleRequestBody` adds named objects to the `SolrQueryResponse`, those objects then become directly available to the client code through SolrJ's response object.

Currently `BrowseList` and `BrowseItem` are wrappers around an `ArrayList<Map<String, Object>>` and a `HashMap<String, Object>` respectively, and have a special `.asMap()` method that just returns those objects. To me that makes it a little confusing to introspect the code and to write tests about content. I'm seriously considering reimplementing `BrowseList` and `BrowseItem` as subclasses of Solr's `NameList` class, or possibly as typed subclasses of `ArrayList` and `HashMap`. That way when I'm unpacking the result and examining the contents, I can cast those objects to the already-familiar `BrowseList` and `BrowseItem` and then it's clear to me and to the reader what I'm working with. And they also easily implement `Serializable` and, depending on the choice, `Iterable`.

What do you think? Does that sound reasonable? Overambitious? Or maybe you see some hidden flaw in the approach? Sanity checks welcome!

marktriggs commented:
@todolson That sounds reasonable to me.  I would have just used ArrayList/HashMap because they were familiar and serialized in the way I wanted at the time, but `NamedList` would fit the bill too I think.

todolson commented:
@marktriggs And that was not in any way intended as a criticism. It didn't occur to me until I was digging through some of the SolrJ code looking for models for how to implement the response class. And then I had to do a lot of experimenting and introspection to figure out what was being returned because the Solr and SolrJ javadoc is not always transparent. I also think some relevant things in Java may have changed since you wrote it. 
(Besides, I always find it easier to edit an existing text than to have to write it in the first place.)

marktriggs commented:
@todolson Heheh no criticism received either, and yeah, SolrJ wasn't on my radar at all at the time, so if you find a better approach I'm all for it :)

todolson commented:
@demiankatz @marktriggs I could use a little advice again as I'm converting `BrowseItem` into a `Collection`.

As a preliminary, it seems like there is no need for `BrowseItem` to keep its contents in a specific order as the browse display code seems to treat it as an object or array. So that's happy.

It seems we want some clear idea of what fields are legit within `BrowseItem`. Currently that is achieved by having separate fields within the class, those fields have defaults and are assigned or added to directly, and then a `HashMap` is produced on demand from those fields:

```java
    // Map as produced today
    public Map<String, Object> asMap()
    {
        Map<String, Object> result = new HashMap<> ();

        result.put("sort_key", sort_key);
        result.put("heading", heading);
        result.put("seeAlso", seeAlso);
        result.put("useInstead", useInstead);
        result.put("note", note);
        result.put("count", new Integer(count));
        result.put("ids", ids);
        result.put("fields", fields);

        return result;
    }
```
One option would be to provide setter methods, so it sort of looks like fields but internally the values are stored in a `HashMap`:
```java
    // Option 1: setter method (or could call them putXXX)
    public void setSeeAlso(List<String> seeAlso) {
        this.put("seeAlso", seeAlso);
    }
    
    public void setUseInstead(List<String> useInstead) {
        this.put("useInstead", useInstead);
    }
```
This still lets us not mess with a wrapper class, and we just use discipline to say "never call `.put()` directly". Assignments would look like object assignments, though the class is really a `Collection`, and the setter (or putter methods if we change naming conventions) document what fields are available:
```java
BrowseItem item = new BrowseItem(sort_key, heading);
item.setNote("This is a note");
```
A second option would be to override the `.put()` method and check that the key names come from a fixed list, something like:
```java
    // Option 2: override put() to restrict possible keys
    public Object put(String key, Object value) {
        if (!(key == "sort_key" || key == "heading" || key == "seeAlso" ||
                key == "useInstead" || key == "note" || key == "count" ||
                key == "ids" || key == "field")) {
            throw new java.lang.IllegalArgumentException("Illegal key: " + key);
        }
        return super.put(key, value);
    }
```
This would be fewer helper methods and would look more like a `Collection` in use, but the caller would need to pay attention to the documentation as the errors would be a run time.

I could try using an enum or maybe a "valid keys" class, but really want to be certain it's easy to consume in testing. 

I'm not entirely happy with either of these, but they would allow `BrowseItem` objects to be represented directly in the results. Does anyone have either a practical or an aesthetic position on this?

marktriggs commented:
My preference would be for the first option--adding helper methods to make setting the right properties more convenient.  I don't see a huge amount of benefit in trying to strictly nail down an interface (with lists of valid keys and that kind of thing): if someone calls `put` directly it's not the end of the world, and the cost of fixing probably wouldn't be huge?

Would a third option be to replace `asMap` with `asNamedList` and just populate and return a `NamedList` instead of a `Map`?  Suggesting this largely out of ignorance, but just thought I'd check.

todolson commented:
As it happens, over the last couple days I've refactored `BrowseItem` as a subclass of `HashMap`, Option 1, to see what that's like. I also implemented the `put()` override, and agree it's not really worth the trouble and should drop it. There are also tests for most of the public methods.

Changing `asMap()` to `asNamedList()` could be an option, but does not achieve my goal, to simply cast objects in the returned results back to `BrowseItem` and then manipulate them in the same way they were constructed. 

I do wish I understood more about how Solr does its internal processing and serializing, maybe I could get away with a straight object with fields. I don't really know that being a `Map` (or maybe just implementing `Iterable<T>`?) is required, just that it works. A next step could maybe be trying to simplify further, or chasing more through the Solr implementation.

marktriggs commented:
Ah, gotcha, that makes sense.

It looks like there are a bunch of serialization options, but straight object would require some work.  It seems to run through this sequence of type checks:

https://github.com/apache/lucene-solr/blob/c97f27b06c1d7c250e9596a9bc7bf5ca11ef6ad3/solr/core/src/java/org/apache/solr/response/TextResponseWriter.java#L120

Then falls through to this one:

https://github.com/apache/lucene-solr/blob/25bca6f16513fda0bdd2ab670633bac26dbf5d6e/solr/solrj/src/java/org/apache/solr/common/util/TextWriter.java#L49

I think if your object implemented WriteableValue, it would get asked to serialize itself by calling the underlying methods on the TextResponseWriter (`writeStr`, `writeInt`, etc.), but that's a fair bit more work.  And even then, it's not clear how you would get a `BrowseItem` back out the other side without also writing a corresponding parser...

todolson commented:
Ah, yes, far from straightforward, which seems to be expected. I had looked earlier ate `JSONResponseWriter` and it was more complicated. It does seem like subclassing `HashMap` and similar is the most straightforward. Thank you for looking at these.

todolson commented:
@marktriggs, as I'm writing some tests I notice that the constructors for `BibDB` and `AuthDB` throw `Exception`, but I don't see any reason they should. They just make field assignments. Do you think it's safe to remove `throws` from the method definitions?

marktriggs commented:
Yep, I think that's safe to remove.  Probably just some legacy weirdness :)

-- 
Mark Triggs
<mark@dishevelled.net>

todolson commented:
@marktriggs I'm trying to write some tests for `BibDB`, particularly `BibDB.recordCount()`, but I'm getting it wrong and not getting a match. Do you see what I'm doing wrong in `BibDBTest.testRecordCount()`?

If I understand `BibDB` correctly, I instantiate it with the an `IndexSearcher` for the  core and the name of the one field in that core which this `BibDB` object will search. Is that correct?

marktriggs commented:
@todolson Yep, that's right.  How is the test data getting loaded?  My first check would be that `searcher.getIndexReader().numDocs()` reported that the index had something in it...

todolson commented:
@marktriggs There's an ant task, `setup-test-cores`. It fires up Solr according to where you set `vufind.home`, imports records, then generates the alphabetical indexes, and then shuts down Solr. The big `test` task has it as a dependency. The big `test` task is the same as:
```
ant -Dvufind.dir=/usr/local/vufind build setup-test-cores main
```
I can fire up Solr and confirm through the dashboard that the data is there. But you are making me think freshly. Here's the problem:

To test hit count on titles in the default browse configuration, (1) must search the biblio core on `title_fullStr`, and (2) use the exact string, not lower case version.

Thanks!

(And that could raise something for a future revision, how we do or do not normalize title browse.)

marktriggs commented:
@todolson Ah, great!  Glad to be a sounding board :)

todolson commented:
@marktriggs @demiankatz I'm looking for a little advice, or maybe a sanity check.
`BibDB.matchingIDs()` now does more than return matching IDs, it also now returns the extra fields. It returns a `Map<String, List<Collection<String>>>`,  where the key is the field name. The comments say "values are `List<Collection>` because some extra fields may be multi-valued", which I'm not certain why `List<Collection<String>>` instead of just accumulating a `List<String>`. It has the property of preserving a `Collection<String>` for each returned document that matches the heading. If there is no need to preserve this structure, we can get a quick if minor improvement by accumulating the values straight into a `Collection` rather than a `List<Collection>`.

I am murky on the purpose of preserving this structure. I _think_ that is preserved all the way to the JSON output that the PHP code will consume, so in theory one could march down the lists in lock-step and know which field came from each bib ID. Can either of you confirm this, and also say whether there is a need to preserve that nested structure?

My sense is that it's time to reconsider `BibDB.matchingIDs()`. It marches through all of the result set accumulating all of the fields, not just IDs. Is that still the best way to get our data? Maybe. At the very least, we could use a better method name, like `matchingFields()` or something. But I do think there's something about how this method has expanded beyond just getting the matching IDs and the need to optimize it in the face of large hit counts that suggests some rethinking.

The simplest thing would be for `maxBibListSize` to change names to `maxBibsChecked` or somesuch and only return `fields` from N records, where N = `maxBibsChecked`, maybe set a default of 100. We might not get all of the values for some of the fields, but it would short-circuit marching through 3,000 records for a single topic heading.

What do you guys think? First, does the analysis of the return value seem correct, that the structure can perhaps be flattened? 

Second, do you have thoughts on how to reconsider `BibDB.matchingIDs()`? Does the above make sense? Or might you have another thought how to optimize and maybe even simplify the code? 

(Aside: I'm a little tempted to rethink `BibDB` so that an instance is not tied to a field, but so the field name and heading are both passed in to `recordCount()` and `matchingIDs()`, so we search on field/value pairs and that is clear when calling the methods, rather than remember how the object was created. Uncertain on cost/value, but it fits my brain better.)

demiankatz commented:
@todolson, I'll defer to @marktriggs on opinions about the internals since I am not familiar enough to have a strong sense of how things work (though I can review code and form an opinion if you need me to). In any case, what you say makes sense in the abstract, but I don't have enough tangible knowledge to comment on details.

I can address the VuFind side of the transaction, though. There's only one place where VuFind fetches data from the browse handler:

https://github.com/vufind-org/vufind/blob/master/module/VuFindSearch/src/VuFindSearch/Backend/Solr/Backend.php#L292

...and this is essentially only used in two places:

1. The AlphaBrowse channel provider (https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/ChannelProvider/AlphaBrowse.php), which only cares about IDs and titles.

2. The alphabrowse/home.phtml template, where the data is rendered (https://github.com/vufind-org/vufind/blob/master/themes/bootstrap3/templates/alphabrowse/home.phtml). Line 121 is where the "extra fields" stuff gets processed.

At a glance, both places seem to be doing work to simplify the data structure coming back from the browse handler (the `[0][0]` of the channel provider seems kind of awkward, and I'm not sure why all this array merging is necessary in the home.phtml template). Again, I didn't write most of this code and haven't worked with it in a while, so I may be forgetting some reason why it is the way it is... but hopefully looking at the other side of the transaction may shed some light on your questions.

If this is not helpful at all, please let me know and I'll give it a closer look when time permits.

todolson commented:
@demiankatz That actually sounds good and helps bring a few things into focus. We do have some changes then to both the configuration of the browse handler and how it returns results, and I think it is time to firm up our strategy for how to deal with the transition. Which changes do we make hard with a version, and which do we implement such that there is a period of transition?

We can use this posting to capture the list of interface and config changes. I'll come back and edit the post if we need to adjust this list.

## Interface and configuration changes for `browse-handler`

1. Currently the UI code relies on bib IDs for creating the links from the browse list to the resulting searches, using them directly when hitcount is <=5. We would like to transition to always using the heading string for building that search.
2. In the VuFind config, we have an `AlphaBrowse_Extras` section to configure extra fields that we want to display. Will this need to change?
3. The `extra` parameter in the browse handler URL is the current way we ask for fields in addition to IDs. We will move to a 
4. For each browse item, record IDs are returned in an `ids` array and other fields are returned in `extras` array. These will be combined in a `fields` array.
5. Within these arrays, the `extras` fields are lists of collection. Under `fields` we would like to flatten this so each field has a list of values.
6. We want to implement a parameter to put an upper bound on the number of records consulted for any BrowseItem.

This highlights the dependencies between the front end and back end. Do we need to support both the old and new version of all back-end properties in parallel during a transition period? And what should we target for VuFind 5.1?

demiankatz commented:
@todolson, thanks for the summary list. As of this writing, there is an incomplete sentence in point 3; not sure what you were going to say there.

As far as timing goes, we've set a February release date for VuFind 5.1, and as a point release, that should not include any backward-breaking changes, so it probably makes sense to aim for VuFind 6.0 with all of this, if a late Summer, 2019 deadline seems reasonable. Of course, if there are any small, simple optimizations that can be done without changing the public interface, I wouldn't turn them down for 5.1... but given the timeline, that doesn't seem especially likely.

In any case, I don't think we need to worry too much about backward compatibility if we go for a 6.0 target -- we can probably change the browse handler code and VuFind code as a single commit and call it a day. Since the repo is tagged to show which browse handler was distributed with which VuFind release, people have a relatively easy way of fetching historical versions if they need the old interface.

todolson commented:
@demiankatz The fragment of point 3 is fixed.

I think we could at least implement an upper limit in time for 5.1, maybe set up for some of the other changes. I need to get something into place in any event, as we're having problems when lists contain items with large hit counts.

todolson commented:
Maybe one thing we could do on the PHP side in 5.1 would be to stop relying on the bib numbers in the browse list when number of hits >=5, and instead always rely on the heading string. Or maybe turn 5 into a configurable N, which could have a default of 5 or maybe a default of 0. That would give us a little time to confirm this is fine on the PHP side and still be easy to back out if some problem emerges.

todolson commented:
(and that close/reopen was just an erroneous click on the trackpad.)

demiankatz commented:
@todolson, I agree that would be a reasonable place to start... and I don't think there's much value in making the value configurable as opposed to simply dropping the ID-based behavior unless we're really concerned that things don't work right without the IDs... and I have no reason to believe that is a problem. 

How would you like to proceed? Do you need me to carve out some time to make the change? I can probably do that next week at some point if you're not able to get there sooner.

todolson commented:
The code currently loops through all of the results not only for the IDs, but also for any of the extra fields. As it happens, we do not look for extra fields when displaying subject browse, where we see the timeout problem. So I could just code it so that if no fields are asked for, no searching is done. I'll look at the code and see which approach looks more readable/maintainable.

On dropping the ID-based behavior, it would be a huge help if you could take that. I'm not a PHP guy, so while I can make tweaks like that it's kind of costly.

demiankatz commented:
@todolson, sounds good! I'll put together a PR for the browse changes when time permits and link it here. Regarding your searching optimization, once you've figured out what's best, we should discuss when/what to merge so that we can make a 5.1 release quickly but also not lose track of the other work in progress.

todolson commented:
The checklist as the top of PR is now revised.

demiankatz commented:
Looks good, @todolson. Let me know if you need any action on my part at this point!

todolson commented:
@demiankatz I've done the initial revert to using `extras`, please try the VuFind tests to be certain that's correct. That will ensure a solid starting point.

demiankatz commented:
@todolson, I killed two birds with one stone, testing the current work here in combination with https://github.com/vufind-org/vufind/pull/1294. All tests passed, including my own manual inspection of the functionality outside of the automated suite. I've gone ahead and merged https://github.com/vufind-org/vufind/pull/1294 to master. Let me know if/when this is ready to go too!

todolson commented:
@marktriggs @demiankatz I'm working on round-tripping the `BrowseList` and `BrowseItem` objects so I can properly reconstitute them from the response. I have a couple questions about `BrowseList.totalCount`.
1. First, what is this value? This value passes through `HeadingsSlice` after being built up in `HeadingsDB` and , and I'm having a little trouble following the code. Is it the offset in rows from the end of the headings database?
2. Second, is this value used in the UI?

I'm trying to figure out whether I really need to work on full round-tripping, and if so, how.

marktriggs commented:
Hi @todolson,

My reading is that it's the number of headings after the point your browse dropped you.  So if you're browsing a list of 100 headings and your results position you at 75, totalcount would be 25.  It seems to be used for pagination, so if you're at position 75 with a page size of 20, you'll need a next page to see the last 5 headings.

Also, the `HeadingsDB` class appears to also use totalcount to mean something completely different... in that case, it's just the total number of headings.  Just seems to be a cached value to save having to recalculate it, but needlessly confusing :)

demiankatz commented:
@todolson, totalCount does appear to be used in VuFind, and its use appears consistent with Mark's description. See the twenty lines of pagination logic here:

https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/Controller/AlphabrowseController.php#L127

If no further adjustments are needed based on this information, let me know, and I'll repeat the tests when I get back to the office next week. If all passes, we can merge this and begin a new PR for 6.0.

demiankatz commented:
@todolson, I built the latest code and ran VuFind's test suite, and I'm noticing what I think may be a minor problem.

VuFind includes an integration test that uses the 'ids' value to validate the contents of the browse handler response (e.g. making sure that the stated count matches the actual id count). I was able to rewrite the test to instead turn on 'extras' and use the 'format' field instead... this gets the test passing. However, while looking at the response, I noticed that the handler is always returning an 'ids' extra value with no content. Should this be happening? I assume that if ids are not requested as an extra, there should be no ids element in the extras array of the response.

demiankatz commented:
@todolson, a little follow-up: I was able to rewrite my test (see https://github.com/vufind-org/vufind/commit/4880eba08f11de77acb343993b3d97899cfcd9d5) in a way that works with both the current master browse handler and this PR's version. However, I still think the phantom empty 'ids' element in 'extras' is probably something that needs to be removed. I can get a list of IDs by turning on 'id' in the extras -- but this gives me an element named 'id', not 'ids.' I know we had some discussion about moving the 'ids' element in the past, but I can't think why we would need the plural version of the key with the singular one seems to work as expected with existing logic. Am I forgetting something?

todolson commented:
@demiankatz I am not quickly finding the cause of the `ids` under `extras`. This does point out the `BrowseItem#setIds` and `BrowseItem#getIds` should be deprecated.

Ah, it's in `BrowseRequestHandler#handleRequestBody`, 438-446:
```java
        // If fields parameter is not provided, construct from extras parameter
        // NOTE: As implemented, fields will always contain ids.
        //       Should think whether to do this the other way around, check
        //       extras first, and how transition will work.
        if (fields == null) {
            String extras = p.get("extras");
            fields = extras == null || extras.length() == 0 ?
                     "ids" : ("ids:" + extras);
        }
```
Our old idea was to force `ids` into the `extras`, but did not account for `ids` being not a real field name in the Solr documents. That was stupid. Change forthcoming.

demiankatz commented:
Great, thanks, @todolson. I'm just happy that once again, my time invested in test-writing has proven not to be wasted. :-)

demiankatz commented:
Thanks, @todolson. I'll retest before the end of the day and will most likely be able to merge. One question about the merging: should I simply collapse this whole PR into a single commit, or is it better to take time to try to curate a more informative commit history? If you think it's worth the time, do you have any suggestions about the most useful way of rebasing this?

todolson commented:
@demiankatz Good question. I've also been erratic about when I break out classes into separate files. If we did want do turn this into an informative commit history, I think it might be a matter of first breaking out the the classes as a commit, and then look at how the changes layer on top, breaking those out into the different points outlined in the work description at the head of the PR. That would make the changes easier to follow, but would be some work. If we did that, maybe easiest to use a separate branch, make the changes in a logical order, then trash this branch and PR? Again, splitting out those classes really complicates rebasing.

I sort of think that complexity may be more than it is worth. I'd be willing to do some of the work if you think that history is worth having. The heavy lifting is already done.

demiankatz commented:
@todolson, I'm inclined to agree that while that history would be nice in an ideal world, it may not be worth the effort, given that there's still further work to be done. If I merge and collapse the history, it will still include a reference to this PR, which will show the work in smaller chunks... so somebody really needing to dig into what happened will still have something to go on.

If you want to invest some time in improving clarity, maybe a better use of it would be to create a PR to split all the classes to files without changing anything. We can merge that before starting the 6.0 work, and it will make those changes significantly easier to read when they come. What do you think?
